File: 131_ImplementTrieII.cpp
/*
Problem: Implement Trie II (With Count)
Description: Implement trie with insert, countWordsEqualTo, countWordsStartingWith,
             and erase operations.

Approach: Each node stores count of words ending here and count of words passing through.
*/

#include <bits/stdc++.h>
using namespace std;

class TrieNode {
public:
    TrieNode* children[26];
    int wordCount;     // Count of words ending at this node
    int prefixCount;   // Count of words passing through this node
    
    TrieNode() {
        for (int i = 0; i < 26; i++)
            children[i] = nullptr;
        wordCount = 0;
        prefixCount = 0;
    }
};

class Trie {
    TrieNode* root;
    
public:
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* node = root;
        for (char c : word) {
            int idx = c - 'a';
            if (!node->children[idx])
                node->children[idx] = new TrieNode();
            node = node->children[idx];
            node->prefixCount++;
        }
        node->wordCount++;
    }
    
    int countWordsEqualTo(string word) {
        TrieNode* node = root;
        for (char c : word) {
            int idx = c - 'a';
            if (!node->children[idx])
                return 0;
            node = node->children[idx];
        }
        return node->wordCount;
    }
    
    int countWordsStartingWith(string prefix) {
        TrieNode* node = root;
        for (char c : prefix) {
            int idx = c - 'a';
            if (!node->children[idx])
                return 0;
            node = node->children[idx];
        }
        return node->prefixCount;
    }
    
    void erase(string word) {
        TrieNode* node = root;
        for (char c : word) {
            int idx = c - 'a';
            if (!node->children[idx])
                return;
            node = node->children[idx];
            node->prefixCount--;
        }
        node->wordCount--;
    }
};

int main() {
    Trie trie;
    
    trie.insert("apple");
    trie.insert("apple");
    trie.insert("app");
    
    cout << trie.countWordsEqualTo("apple") << endl;        // Output: 2
    cout << trie.countWordsStartingWith("app") << endl;     // Output: 3
    
    trie.erase("apple");
    cout << trie.countWordsEqualTo("apple") << endl;        // Output: 1
    cout << trie.countWordsStartingWith("app") << endl;     // Output: 2
    
    return 0;
}

/*
Time Complexity: O(m) for all operations
Space Complexity: O(alphabet_size * n * m)
*/

File: 131_ImplementTrieII.py
"""
Problem: Implement Trie II (With Count)
Description: Implement trie with insert, countWordsEqualTo, countWordsStartingWith,
             and erase operations.

Approach: Each node stores count of words ending here and count of words passing through.
"""

class TrieNode:
    def __init__(self):
        self.children = {}
        self.wordCount = 0     # Count of words ending at this node
        self.prefixCount = 0   # Count of words passing through this node

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
            node.prefixCount += 1
        node.wordCount += 1
    
    def countWordsEqualTo(self, word: str) -> int:
        node = self.root
        for c in word:
            if c not in node.children:
                return 0
            node = node.children[c]
        return node.wordCount
    
    def countWordsStartingWith(self, prefix: str) -> int:
        node = self.root
        for c in prefix:
            if c not in node.children:
                return 0
            node = node.children[c]
        return node.prefixCount
    
    def erase(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node.children:
                return
            node = node.children[c]
            node.prefixCount -= 1
        node.wordCount -= 1

def main():
    trie = Trie()
    
    trie.insert("apple")
    trie.insert("apple")
    trie.insert("app")
    
    print(trie.countWordsEqualTo("apple"))        # Output: 2
    print(trie.countWordsStartingWith("app"))     # Output: 3
    
    trie.erase("apple")
    print(trie.countWordsEqualTo("apple"))        # Output: 1
    print(trie.countWordsStartingWith("app"))     # Output: 2

if __name__ == "__main__":
    main()

"""
Time Complexity: O(m) for all operations
Space Complexity: O(alphabet_size * n * m)
"""
